### PR TITLE
Operator tools

### PR DESCRIPTION
**Pull request by joshcombes**
This is a first pass. Note these things

0. major module re-org
1. fix to plotting module
2. update random operators notebook
3. update superoperator tools notebook
4. type checking in random ops
5. new module to check plain old operators are unitary etc

Not sure that I am happy with the current names or organization. I liked having a big monolithic superoperators module. The imports via `__init__` is awkward, try it in a notebook.

Feel free to change and make commits on this branch as you see fit.

### CODE REVIEW

**Comment by joshcombes** [comment]
@kylegulshen i cant figure the failure out. the t2 notebook runs fine locally.

**Comment by joshcombes** [comment]
ok now they work ...

File: forest/benchmarking/operator_tools/random_operators.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    rho_red = partial_trace(rho, [0], [dim, dim])
+    # Note that Eqn. 8 of [RQO] uses a *row* stacking convention so in that case we would write
+    # Q = np.kron(np.eye(D), sqrtm(la.inv(rho_red)))
+    # But as we use column stacking we need:
+    Q = np.kron(sqrtm(la.inv(rho_red)), np.eye(dim))
+    Z = Q @ rho @ Q
+    return Z
+
+
+def permute_tensor_factors(dims: Union[int, List[int]], perm: List[int]) -> np.ndarray:
```
Comment (kylegulshen) [inline forest/benchmarking/operator_tools/random_operators.py]: Should we move this to somewhere else? It doesn't really fit here.

File: forest/benchmarking/plotting/state_process.py [hunk 2]
```diff
@@ -110,11 +110,12 @@ def plot_pauli_transfer_matrix(ptransfermatrix, ax, labels=None, title='', fonts
     ticklabs = cb.ax.get_yticklabels()
     cb.ax.set_yticklabels(ticklabs, ha='right')
     cb.ax.yaxis.set_tick_params(pad=35)
+    cb.draw_all()
     ax.set_xticks(range(dim_squared))
     ax.set_xlabel("Input Pauli Operator", fontsize=fontsizes)
     ax.set_yticks(range(dim_squared))
     ax.set_ylabel("Output Pauli Operator", fontsize=fontsizes)
-    ax.set_title(title, fontsize= int(np.floor(1.2*fontsizes)))
+    ax.set_title(title, fontsize= int(np.floor(1.2*fontsizes)), pad=15)
```
Comment (kylegulshen) [inline forest/benchmarking/plotting/state_process.py]: This pad causes an `AttributeError: Unkown property pad`  for me when running random_operators.ipynb

Comment (joshcombes) [reply forest/benchmarking/plotting/state_process.py]: This is probably a matplotlib version problem. Perhaps we should upgrade the requirements.

File: forest/benchmarking/operator_tools/random_operators.py (hunk 1 above)
Comment (joshcombes) [reply forest/benchmarking/operator_tools/random_operators.py]: Ok lets make a PR to put this somewhere else.

**Comment by joshcombes** [comment]
partial close of https://github.com/rigetti/forest-benchmarking/issues/98

**Pull request merged**